### PR TITLE
log_backup: implement checkpoint v3 advancer

### PR DESCRIPTION
**Pull request by YuJuncen**
### What problem does this PR solve?

Issue Number: close #(TBD)

Problem Summary:
We need a central node for advancing the global checkpoint.

### What is changed and how it works?
This PR:
- Added a new package `streamhelper`, which contains the advancer type and client type.
- The "advancer" is a central node, which would poll the gRPC service `LogBackup` for the latest checkpoint of each region, collect them and advancing the global checkpoint.
- The "advancer" can be enabled by either:
  - Use the `br log advancer` command.
  - Set the tidb config `log-backup.enabled` to true.

### Check List

Tests 

- [x] Unit test
- [ ] Integration test
- [x] Manual test (add detailed scripts or steps below)
- [ ] No code

### Release note

Please refer to [Release Notes Language Style Guide](https://pingcap.github.io/tidb-dev-guide/contribute-to-tidb/release-notes-style-guide.html) to write a quality release note.

```release-note
None
```

### CODE REVIEW

File: br/pkg/streamhelper/advancer.go [hunk 1]
```diff
@@ -0,0 +1,428 @@
...
+	return nil
+}
+
+func (c *CheckpointAdvancer) advanceCheckpointBy(ctx context.Context, getCheckpoint func() (uint64, error)) error {
+	start := time.Now()
+	cp, err := getCheckpoint()
+	if err != nil {
+		return err
+	}
+	if cp < c.lastCheckpoint {
```
Comment (joccau) [inline br/pkg/streamhelper/advancer.go:348]:
```suggestion
	if cp <= c.lastCheckpoint {
```

File: br/pkg/streamhelper/collector.go [hunk 2]
```diff
@@ -0,0 +1,284 @@
...
+		sc := StoreCheckpoints{
+			HasCheckpoint:    c.checkpoint != 0,
+			Checkpoint:       c.checkpoint,
+			FailureSubranges: c.inconsistent,
+		}
+		return sc, nil
+	}
+}
+
+func (c *storeCollector) flush(ctx context.Context) error {
```
Comment (joccau) [inline br/pkg/streamhelper/collector.go:176]: the name `flush` is not proper, I consider, suggest changing it. Maybe `runBatchRequest` ?

File: br/pkg/streamhelper/collector.go [hunk 3]
```diff
@@ -0,0 +1,284 @@
...
+	if err != nil {
+		return err
+	}
+	cps, err := cli.GetLastFlushTSOfRegion(ctx, &c.currentRequest)
+	if err != nil {
+		return err
+	}
+	metrics.GetCheckpointBatchSize.WithLabelValues("checkpoint").Observe(float64(len(c.currentRequest.GetRegions())))
+	c.currentRequest = logbackup.GetLastFlushTSOfRegionRequest{}
+	for _, checkpoint := range cps.Checkpoints {
```
Comment (joccau) [inline br/pkg/streamhelper/collector.go:188]: Can it remove region KeyRange information from regionMap to safe memory if it succeeds get checkpoints for a specify region ?

File: br/pkg/streamhelper/collector.go [hunk 4]
```diff
@@ -0,0 +1,284 @@
...
+func (c *clusterCollector) collectRegion(r RegionWithLeader) error {
+	c.mu.Lock()
+	defer c.mu.Unlock()
+	if c.masterCtx.Err() != nil {
+		return nil
+	}
+
+	if r.Leader.GetStoreId() == 0 {
+		log.Warn("there is regions without leader", zap.Uint64("region", r.Region.GetId()))
+		c.noLeaders = append(c.noLeaders, kv.KeyRange{StartKey: r.Region.StartKey, EndKey: r.Region.EndKey})
```
Comment (joccau) [inline br/pkg/streamhelper/collector.go:248]: need return

File: br/pkg/streamhelper/advancer.go [hunk 5]
```diff
@@ -0,0 +1,428 @@
...
+		thisRun = nextRun
+		nextRun = nil
+		log.Debug("backoffing with subranges", zap.Int("subranges", len(thisRun)))
+		time.Sleep(c.cfg.BackoffTime)
+	}
+}
+
+func CollapseRanges(length int, getRange func(int) kv.KeyRange) []kv.KeyRange {
+	frs := make([]kv.KeyRange, 0, length)
+	for i := 0; i < length; i++ {
```
Comment (joccau) [inline br/pkg/streamhelper/advancer.go:226]: Why not set `ranges []kv.KeyRange` as input arguements directly, but a closure ?

File: br/pkg/streamhelper/collector.go (hunk 4 above)
Comment (3pointer) [reply br/pkg/streamhelper/collector.go:248]: and another thing is `noLeaders` seems always increase and never clear.

File: br/pkg/streamhelper/collector.go [hunk 6]
```diff
@@ -0,0 +1,284 @@
...
+type storeCollector struct {
+	storeID   uint64
+	batchSize int
+
+	service LogBackupService
+
+	input chan RegionWithLeader
+	// the oneshot error reporter.
+	// should only be used to report the sole call of "recvLoop".
+	errMessenger chan error
```
Comment (3pointer) [inline br/pkg/streamhelper/collector.go:39]: consider use an atomic value of error.

**Comment by 3pointer** [comment]
can we add a check that if this round checkpoint advance not finished, then do not start next round.

**Review by 3pointer — APPROVED** [review]

File: br/pkg/streamhelper/collector.go (hunk 3 above)
Comment (YuJuncen) [reply br/pkg/streamhelper/collector.go:193]: We should keep the key range, this is fundamental for the consistency of checkpoint; if the key ranges weren't recorded, we must scan the full key range of the task for getting a safe checkpoint.

File: br/pkg/streamhelper/advancer.go (hunk 5 above)
Comment (YuJuncen) [reply br/pkg/streamhelper/advancer.go:296]: At the very beginning, this function may accept `[]Region`, which contains a key range but cannot be passed directly. This signature can handle that condition. And for now I think keep this is still OK.

**Comment by YuJuncen** [comment]
`TestGetTS` 🤔
/run-tests

**Review by joccau — APPROVED** [review]
LGTM

**Pull request merged**